### PR TITLE
Add SubnetPort CRD

### PR DESCRIPTION
**Pull request by heypnus**

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -6,16 +6,20 @@ require (
 	github.com/agiledragon/gomonkey v2.0.2+incompatible
 	github.com/agiledragon/gomonkey/v2 v2.4.0
 	github.com/deckarep/golang-set v1.8.0
+	github.com/go-logr/logr v1.2.0
```
Comment (jwsui) [inline go.mod:9]: Do we need to add these dependencies before adding controller code?

**Review by jwsui — COMMENTED** [review]
Any update to zz_generated.deepcopy.go?

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+	Message string `json:"message,omitempty"`
+}
+
+// SubnetPortStatusCondition is an aspect of operator state.
+type SubnetPortStatusCondition string
+
+type SubnetPortIPAddress struct {
+	Gateway string `json:"gateway,omitempty"`
+	IP string `json:"ip,omitempty"`
+	SubnetMask string `json:"subnetMask,omitempty"`
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/subnetport_types.go:54]: Could we add the below validation for subnetMask?
// SubnetMask defines an IPv4 range in the form <ip>/<bits>.
// +kubebuilder:validation:Pattern="^(?:(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\\.){3}(?:25[0-5]|2[0-4][0-9]|1[0-9][0-9]|[1-9][0-9]|[0-9])\\/((3[0-2])|([1-2][0-9])|[1-9])$"

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+
+// SubnetPortSpec defines the desired state of SubnetPort
+type SubnetPortSpec struct {
+	// Subnet defines the attachement of the SubnetPort
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/subnetport_types.go:14]: And "." at the end of sentence.

File: go.mod (hunk 1 above)
Comment (heypnus) [reply go.mod:9]: Done. Thanks for the reminder.

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 3 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go:14]: Done

**Comment by heypnus** [comment]
> Any update to zz_generated.deepcopy.go?

Done. Thanks for the reminder.

File: build/yaml/crd/nsx.vmware.com_subnetports.yaml [hunk 4]
```diff
@@ -0,0 +1,110 @@
...
+          spec:
+            description: SubnetPortSpec defines the desired state of SubnetPort
+            properties:
+              VirtualMachine:
+                description: VirtualMachine labels to realize SecurityPolices (VMSelector)
+                type: string
+              subnet:
+                description: Subnet defines the attachement of the SubnetPort
+                type: string
+              subnetset:
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_subnetports.yaml:45]: Not sure if we should set VM/Subnet/SubnetSet in ownerReference for SubnetPort CR

File: build/yaml/crd/nsx.vmware.com_subnetports.yaml [hunk 5]
```diff
@@ -0,0 +1,110 @@
...
+              submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds'
+            type: string
+          metadata:
+            type: object
+          spec:
+            description: SubnetPortSpec defines the desired state of SubnetPort
+            properties:
+              VirtualMachine:
+                description: VirtualMachine labels to realize SecurityPolices (VMSelector)
+                type: string
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_subnetports.yaml:41]: Do you mean the value of VirtualMachine is label?

File: build/yaml/crd/nsx.vmware.com_subnetports.yaml [hunk 6]
```diff
@@ -0,0 +1,110 @@
...
+                    status:
+                      description: Status shows the status of condition, one of True
+                        or False.
+                      type: string
+                    type:
+                      description: Type defines the type of condition.
+                      type: string
+                  type: object
+                type: array
+              interfaceID:
```
Comment (dantingl) [inline build/yaml/crd/nsx.vmware.com_subnetports.yaml:73]: interfaceID should be the VIF ID instead of LSP ID, right?

Comment (heypnus) [reply build/yaml/crd/nsx.vmware.com_subnetports.yaml:90]: Yes you are right... Fixed.

File: build/yaml/crd/nsx.vmware.com_subnetports.yaml (hunk 4 above)
Comment (heypnus) [reply build/yaml/crd/nsx.vmware.com_subnetports.yaml:62]: I'm asking to add VM in the `ownerReferences` like what virtualnetworkinterface already did.. And I think we'd better not add multiple owners in the `ownerReferences`, which may introduce unnecessary complexity.

File: PROJECT [hunk 7]
```diff
@@ -20,4 +20,12 @@ resources:
   kind: SecurityPolicy
   path: github.com/vmware-tanzu/nsx-operator/pkg/api/v1alpha1
   version: v1alpha1
+- api:
```
Comment (jwsui) [inline PROJECT:23]: What is this used for, and is this generated automatically or manually？

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// SubnetPortSpec defines the desired state of SubnetPort.
+type SubnetPortSpec struct {
+	// Subnet defines the attachement of the SubnetPort.
+	Subnet string `json:"subnet,omitempty"`
+	// SubnetSet defines the attachement of the SubnetPort.
+	SubnetSet string `json:"subnetset,omitempty"`
+	// Attachment refers to the virtual machine which the SubnetPort is attached
```
Comment (jwsui) [inline pkg/apis/v1alpha1/subnetport_types.go:17]: Add '.' at the end?

**Review by jwsui — COMMENTED** [review]
Consider add build/yaml/samples if necessary?

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+// SubnetPortStatus defines the observed state of SubnetPort.
+type SubnetPortStatus struct {
+	// Conditions describes current state of SubnetPort.
+	Conditions []SubnetPortCondition `json:"conditions,omitempty"`
+	// InterfaceID describes the attachment VIF ID owned by the SubnetPort in NSX-T.
+	InterfaceID string `json:"interfaceID,omitempty"`
+	// IPAddresses describes the IP addresses of the SubnetPort.
+	IPAddresses []SubnetPortIPAddress `json:"ipAddresses,omitempty"`
+	// MacAddress describes the MAC address of the SubnetPort.
+	MacAddress string `json:"macAddress,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/v1alpha1/subnetport_types.go:30]: Mac -> MAC

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+	// Attachment refers to the virtual machine which the SubnetPort is attached.
+	Attachment corev1.TypedLocalObjectReference `json:"attachment,omitempty"`
+}
+
+// SubnetPortStatus defines the observed state of SubnetPort.
+type SubnetPortStatus struct {
+	// Conditions describes current state of SubnetPort.
+	Conditions []SubnetPortCondition `json:"conditions,omitempty"`
+	// InterfaceID describes the attachment VIF ID owned by the SubnetPort in NSX-T.
+	InterfaceID string `json:"interfaceID,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/v1alpha1/subnetport_types.go:26]: What is the terminology in NSX-T API? I guess we should use either VIFID to be consistent with NSX-T API, or ExternalID to be consistent with VAPI. @lxiaopei

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+	// IPAddresses describes the IP addresses of the SubnetPort.
+	IPAddresses []SubnetPortIPAddress `json:"ipAddresses,omitempty"`
+	// MacAddress describes the MAC address of the SubnetPort.
+	MacAddress string `json:"macAddress,omitempty"`
+	// ProviderStatus defines the NSX-T resource provider status.
+	ProviderStatus SubnetPortProviderStatus `json:"providerStatus,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/v1alpha1/subnetport_types.go:32]: I feel we can just put LogicalSwitchID here, and need not a "SubnetPortProviderStatus" struct. We do not have multiple providers right.

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+)
+
+// SubnetPortSpec defines the desired state of SubnetPort.
+type SubnetPortSpec struct {
+	// Subnet defines the parent Subnet name of the SubnetPort.
+	Subnet string `json:"subnet,omitempty"`
+	// SubnetSet defines the parent SubnetSet name of the SubnetPort.
+	SubnetSet string `json:"subnetset,omitempty"`
+	// Attachment refers to the virtual machine which the SubnetPort is attached.
+	Attachment corev1.TypedLocalObjectReference `json:"attachment,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/v1alpha1/subnetport_types.go:18]: I thought we plan to name it "OwnerReference"? @heypnus

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 10 above)
Comment (lxiaopei) [reply pkg/apis/v1alpha1/subnetport_types.go:26]: In NSX API, it is the id in Port Attachment. Shall we name it PortAttachmentID or VIFID? Which do you prefer? @jianjuns 
- !Type
  id: PortAttachment
  title: Attachment information on the port
  type: object
  description: Detail information about port attachment
  properties:
    id:
      title: Port attachment ID
      type: string
      description: VIF UUID on NSX Manager. If the attachement type is PARENT, this
        property is required.
      required: false

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 11 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go:32]: make sense, done

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 12 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go]: done

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 13]
```diff
@@ -0,0 +1,64 @@
+/* Copyright © 2021 VMware, Inc. All Rights Reserved.
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/subnetport_types.go:1]: 2021 -> 2022

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 10 above)
Comment (lxiaopei) [reply pkg/apis/v1alpha1/subnetport_types.go:26]: Probably we can change to VIFUUID. @heypnus what do you think?

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+)
+
+// SubnetPortSpec defines the desired state of SubnetPort.
+type SubnetPortSpec struct {
+	// Subnet defines the parent Subnet name of the SubnetPort.
+	Subnet string `json:"subnet,omitempty"`
+	// SubnetSet defines the parent SubnetSet name of the SubnetPort.
+	SubnetSet string `json:"subnetSet,omitempty"`
+	// AttachmentRef refers to the virtual machine which the SubnetPort is attached.
+	AttachmentRef corev1.TypedLocalObjectReference `json:"attachmentRef,omitempty"`
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/subnetport_types.go:18]: I checked corev1.TypedLocalObjectReference is defined as :
       APIGroup *string `json:"apiGroup,omitempty"`
	Kind     *string `json:"kind,omitempty"`
	Name     *string `json:"name,omitempty"`

Is there any difference between apiGroup and apiVersion?

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 15]
```diff
@@ -0,0 +1,64 @@
...
+	// MACAddress describes the MAC address of the SubnetPort.
+	MACAddress string `json:"macAddress,omitempty"`
+	// LogicalSwitchID defines the logical switch ID in NSX-T.
+	LogicalSwitchID string `json:"logicalSwitchID,omitempty"`
+}
+
+type SubnetPortIPAddress struct {
+	Gateway    string `json:"gateway,omitempty"`
+	IP         string `json:"ip,omitempty"`
+	SubnetMask string `json:"subnetMask,omitempty"`
```
Comment (jianjuns) [inline pkg/apis/v1alpha1/subnetport_types.go:38]: Name it "Netmask"?

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 14 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go]: Changed it to `corev1.ObjectReference`

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 10 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go]: Renamed it to `VIFID`, thanks.

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 15 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go]: Done

File: pkg/apis/v1alpha1/subnetport_types.go [hunk 16]
```diff
@@ -0,0 +1,64 @@
...
+	SubnetSet string `json:"subnetSet,omitempty"`
+	// AttachmentRef refers to the virtual machine which the SubnetPort is attached.
+	AttachmentRef corev1.ObjectReference `json:"attachmentRef,omitempty"`
+}
+
+// SubnetPortStatus defines the observed state of SubnetPort.
+type SubnetPortStatus struct {
+	// Conditions describes current state of SubnetPort.
+	Conditions []Condition `json:"conditions,omitempty"`
+	// InterfaceID describes the attachment VIF ID owned by the SubnetPort in NSX-T.
```
Comment (lxiaopei) [inline pkg/apis/v1alpha1/subnetport_types.go:25]: InterfaceID -> VIFID

**Comment by lxiaopei** [comment]
a minor comment, others LGTM.

File: pkg/apis/v1alpha1/subnetport_types.go (hunk 16 above)
Comment (heypnus) [reply pkg/apis/v1alpha1/subnetport_types.go]: Done, thanks!

**Review by lxiaopei — APPROVED** [review]

**Pull request merged**